// @clerk/shared/dist/runtime/error.js
const require_error = require('./error-DyScPmkp.js');

exports.ClerkAPIError = require_error.ClerkAPIError;
exports.ClerkAPIResponseError = require_error.ClerkAPIResponseError;
exports.ClerkError = require_error.ClerkError;
exports.ClerkRuntimeError = require_error.ClerkRuntimeError;
exports.ClerkWebAuthnError = require_error.ClerkWebAuthnError;
exports.EmailLinkError = require_error.EmailLinkError;
exports.EmailLinkErrorCode = require_error.EmailLinkErrorCode;
exports.EmailLinkErrorCodeStatus = require_error.EmailLinkErrorCodeStatus;
exports.buildErrorThrower = require_error.buildErrorThrower;
exports.createClerkGlobalHookError = require_error.createClerkGlobalHookError;
exports.errorToJSON = require_error.errorToJSON;
exports.is4xxError = require_error.is4xxError;
exports.isCaptchaError = require_error.isCaptchaError;
exports.isClerkAPIResponseError = require_error.isClerkAPIResponseError;
exports.isClerkRuntimeError = require_error.isClerkRuntimeError;
exports.isEmailLinkError = require_error.isEmailLinkError;
exports.isKnownError = require_error.isKnownError;
exports.isMetamaskError = require_error.isMetamaskError;
exports.isNetworkError = require_error.isNetworkError;
exports.isPasswordPwnedError = require_error.isPasswordPwnedError;
exports.isReverificationCancelledError = require_error.isReverificationCancelledError;
exports.isUnauthorizedError = require_error.isUnauthorizedError;
exports.isUserLockedError = require_error.isUserLockedError;
exports.parseError = require_error.parseError;
exports.parseErrors = require_error.parseErrors;
